runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,905 @@
+pub struct DisputesTimeSlot {
+	// The order of these matters for `derive(Ord)`.
+	session_index: SessionIndex,
+	candidate_hash: CandidateHash,
+}
+
+impl DisputesTimeSlot {
+	pub fn new(session_index: SessionIndex, candidate_hash: CandidateHash) -> Self {
+		Self { session_index, candidate_hash }
+	}
+}
+
+/// An offence that is filed when a series of validators lost a dispute
+/// about an invalid candidate.
+#[derive(RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq, Eq))]
+pub struct ForInvalidOffence<KeyOwnerIdentification> {
+	/// The size of the validator set in that session.
+	pub validator_set_count: ValidatorSetCount,
+	/// Should be unique per dispute.
+	pub time_slot: DisputesTimeSlot,
+	/// Staking information about the validators that lost the dispute
+	/// needed for slashing.
+	pub offenders: Vec<KeyOwnerIdentification>,
+}
+
+impl<Offender> Offence<Offender> for ForInvalidOffence<Offender>
+where
+	Offender: Clone,
+{
+	const ID: Kind = *b"disputes:invalid";
+
+	type TimeSlot = DisputesTimeSlot;
+
+	fn offenders(&self) -> Vec<Offender> {
+		self.offenders.clone()
+	}
+
+	fn session_index(&self) -> SessionIndex {
+		self.time_slot.session_index
+	}
+
+	fn validator_set_count(&self) -> ValidatorSetCount {
+		self.validator_set_count
+	}
+
+	fn time_slot(&self) -> Self::TimeSlot {
+		self.time_slot.clone()
+	}
+
+	fn disable_strategy(&self) -> DisableStrategy {
+		// The default is `DisableStrategy::DisableWhenSlashed`
+		// which is true for this offence type, but we're being explicit
+		DisableStrategy::Always
+	}
+
+	fn slash_fraction(_offenders: u32, _: ValidatorSetCount) -> Perbill {
+		Perbill::from_percent(100)
+	}
+}
+
+/// An offence that is filed when a series of validators lost a dispute
+/// about an valid candidate.
+#[derive(RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq, Eq))]
+pub struct AgainstValidOffence<KeyOwnerIdentification> {
+	/// The size of the validator set in that session.
+	pub validator_set_count: ValidatorSetCount,
+	/// Should be unique per dispute.
+	pub time_slot: DisputesTimeSlot,
+	/// Staking information about the validators that lost the dispute
+	/// needed for slashing.
+	pub offenders: Vec<KeyOwnerIdentification>,
+}
+
+impl<Offender> Offence<Offender> for AgainstValidOffence<Offender>
+where
+	Offender: Clone,
+{
+	const ID: Kind = *b"disputes:valid::";
+
+	type TimeSlot = DisputesTimeSlot;
+
+	fn offenders(&self) -> Vec<Offender> {
+		self.offenders.clone()
+	}
+
+	fn session_index(&self) -> SessionIndex {
+		self.time_slot.session_index
+	}
+
+	fn validator_set_count(&self) -> ValidatorSetCount {
+		self.validator_set_count
+	}
+
+	fn time_slot(&self) -> Self::TimeSlot {
+		self.time_slot.clone()
+	}
+
+	fn disable_strategy(&self) -> DisableStrategy {
+		DisableStrategy::Never
+	}
+
+	fn slash_fraction(_offenders: u32, _: ValidatorSetCount) -> Perbill {
+		Perbill::from_percent(1)
```

Comment:
Would it make sense to have the slashing parameters in `HostConfiguration` ?

Reply:
I was thinking about it. Currently, this interface requires a static method and we might want to change it to accept `&self`. And add `slash_fraction` to the dispute type on construction reading from the `HostConfiguration`. 

But I decided to start simple.

Reply:
Does it sound like a sane approach to start with constants first and then add these as configurables via `HostConfiguration` (which would also require a runtime migration)?